content/blog/2024-06-17-gsoc-multiboot2.mdx

```diff
@@ -0,0 +1,55 @@
+---
+title: "Multiboot2 Support in Unikraft"
+description: |
+  Multiboot2, an enhanced version of Multiboot, solves some of the limitations of the original protocol, such as the lack of support for 64-bit UEFI systems. It also provides a more accurate and detailed communication during the boostrapping process and offers a standardized tag system for easier configuration.
+publishedDate: 2024-06-17
+tags:
+- GSoC'24
+- Multiboot2
+- Booting
+authors:
+- Maria Pana
+---
+
+<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-generic-badge.svg" align="right" />
+
+## GSoC'24: Multiboot2 Support in Unikraft
+
+### Project overview
+
+When powering on an x86 system, it starts in Real Mode (the CPU is in 16-bit mode, and the memory is limited to 1MB).
+In order to gain more information about the system, it can interact with BIOS, which initially runs self test and initialization routines on hardware.
```

Comment:
Hmm it is not clear who interacts with the BIOS and who wants to get more information about the system.
I would reword this sentence together with the last one to keep it clear :thinking:. I will try to shed some more light on things in the hopes that it will help you when rewording these :smiley:. I know the documentation for these is not always straightforward.

There exist two types of BIOS specifications (that I know of): PnP (Plug and Play) BIOS (aka legacy BIOS) and UEFI BIOS. Both are the same thing really, the difference comes to what they offer as interfaces to the software succeeding them (a bootloader for example) and the steps they take to achieve the state of being able to offer such interfaces (e.g. legacy BIOS interrupt routines vs UEFI System Table).

On x86 systems, regardless of whether PnP or UEFI, they start from a reset vector in Real Mode at a certain hardcoded address (I will let you find its value). From then on they can jump between Protected Mode, Unreal Mode, or 64-bit mode of Long Mode (UEFI 64-bit case) to achieve their goals.

PnP BIOS traditionally after starting in Real Mode, it may jump between Unreal Mode and Protected Mode to initialize hardware and set up resources that it can pass to the succeeding software to describe the systems to it in a manner that is motherboard/chipset/whatever agnostic (e.g. the memory map you saw in Multiboot2 actually originates from a BIOS call) only to then look through the list of bootable devices. The first one it finds that has a valid MBR, whose first few hundred bytes contain the bootstrap code is the one which control is passed to. But right before that, BIOS descends back to Real Mode (if it was not already there, it has full freedom to operate in whatever mode it wants to) as this is how it is agreed upon through the PnP specification, that the MBR bootstrap code starts in Real Mode (that code is also loaded at a very specific address, I will let you find out what that is).

UEFI BIOS similarly to the PnP one may jump between modes but may mostly run in 64-bit mode depending on the implementation (vendor's choice, but 32-bit UEFI is rare) and follows conceptually similar initialization steps, just under different form (think different looking C structs, but the same purpose). Unlike legacy BIOS, it looks for bootable devices with a GPT (or a (P)MBR for backwards compatibility) partitioning scheme, finds the EFI System Partition (FAT filesystem) in the GPT and boots the binary that has a special name (I will leave you to finding out what that name is). It does so while still being in 64-bit (or 32-bit for the 32-bit UEFI we don't care about).

Once loaded by the BIOS, the bootloader makes use of the aforementioned interfaces installed by the BIOS to be able to load the kernel and fill in the boot protocol structures, e.g.:
- the bootloader finds the memory map by using either the [E820 call](<http://www.uruk.org/orig-grub/mem64mb.html>) in the case of legacy BIOS, or [GetMemoryMap](<https://uefi.org/specs/UEFI/2.10/07_Services_Boot_Services.html#efi-boot-services-getmemorymap>) in the case of UEFI systems
- the bootloader is able to load the kernel from storage (e.g. disk) by using [INT 13 call](<https://en.wikipedia.org/wiki/INT_13H>) in the case of legacy BIOS, or [LoadFile](<https://uefi.org/specs/UEFI/2.10/13_Protocols_Media_Access.html#efi-load-file2-protocol-loadfile>) in the case of UEFI

Of course, the bootloader may not use these interfaces at all if it does not want to :smile:. In fact, it is a good thing for a bootloader to use as little BIOS calls as possible (firmware tends to be buggy) and have its own drivers ready that it can control (though while some are avoidable e.g. calls for disk, some are unavoidable, e.g. memory map fetching). For example, Windows Boot Manger is a famously very well written bootloader which is said to have around 7 UEFI calls only (that is very little!), choosing to rely mostly on its own drivers instead.

So the need for a bootloader does not necessarily come from Real Mode being difficult, as it is very easy to jump to 64-bit. Essentially, a BIOS is needed to initialize the very motherboard/vendor specific things only it itself needs to know only to then install some interfaces to make it easy for the software that comes afterwards to discover the resources of the system in a generic manner, so that it does not need to have a driver for every single motherboard out there. A bootloader like GRUB helps here as it can know the BIOS interfaces for the kernel, which rids the kernel of the need to speak any of these. That being said, the bootloader phase is for the comfortability of the kernel, but it can be skipped entirely if the kernel also knows these interfaces (e.g. see [EFI stub](<https://docs.kernel.org/admin-guide/efi-stub.html>))

Reply:
The wording is indeed unclear, but I am glad and grateful for your insight! Thank you! I will be able to provide a much clearer view of the process now :smiley: 

I will push the modification, please let me know if I should go into more specifics with the explanation. Thanks again!